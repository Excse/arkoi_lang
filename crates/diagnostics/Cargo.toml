[package]
name = "diagnostics"
version = "0.1.0"
edition = "2021"

[dependencies]
colored = "2.0.4"
serde = { version = "1.0.188", features = ["derive"] }
strum = "0.25"
strum_macros = "0.25"
