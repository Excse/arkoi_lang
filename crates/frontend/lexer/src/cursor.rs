#[cfg(feature = "serialize")]
use serde::Serialize;

use std::{iter::Peekable, str::CharIndices};

use crate::error::{DidntExpect, EndOfFile, Result};
use diagnostics::{
    file::{FileID, Files},
    positional::{LabelSpan, Span},
};

#[cfg_attr(feature = "serialize", derive(Serialize))]
#[derive(Debug)]
pub struct Cursor<'a> {
    file_id: FileID,
    files: &'a Files,
    #[serde(skip)]
    chars: Peekable<CharIndices<'a>>,
    length: usize,
    start: usize,
}

impl<'a> Cursor<'a> {
    pub fn new(file_id: FileID, files: &'a Files) -> Cursor<'a> {
        let source = files
            .source(file_id)
            .expect("Couldn't get the source of this file.");

        Self {
            file_id,
            files,
            chars: source.char_indices().peekable(),
            length: source.len(),
            start: 0,
        }
    }

    pub fn current_index(&mut self) -> usize {
        self.peek_indexed()
            .map(|(index, _)| index)
            .unwrap_or(self.length)
    }

    pub fn mark_start(&mut self) {
        self.start = self.current_index()
    }

    pub fn as_span(&mut self) -> LabelSpan {
        LabelSpan::new(Span::new(self.start, self.current_index()), self.file_id)
    }

    // TODO: Remove the expect
    pub fn as_str(&mut self) -> &'a str {
        let span = self.as_span();
        self.files
            .slice(self.file_id, &span.span)
            .expect("Couldn't slice the source")
    }

    pub fn peek_indexed(&mut self) -> Option<(usize, char)> {
        self.chars.peek().copied()
    }

    pub fn peek(&mut self) -> Option<char> {
        self.peek_indexed().map(|(_, char)| char)
    }

    pub fn try_consume(&mut self) -> Option<char> {
        let char = self.chars.next().map(|(_, char)| char)?;
        Some(char)
    }

    pub fn try_eat(&mut self, expected: char) -> Result<char> {
        match self.peek_indexed() {
            Some((_, char)) if char == expected => Ok(self.try_consume().unwrap()),
            Some((index, char)) => Err(DidntExpect::new(
                char,
                LabelSpan::new(Span::single(index), self.file_id),
                expected,
            )
            .into()),
            None => Err(EndOfFile.into()),
        }
    }

    pub fn eat_if<F>(&mut self, predicate: F, message: &'static str) -> Result<char>
    where
        F: FnOnce(char) -> bool,
    {
        match self.peek_indexed() {
            Some((_, char)) if predicate(char) => Ok(self.try_consume().unwrap()),
            Some((index, char)) => Err(DidntExpect::new(
                char,
                LabelSpan::new(Span::single(index), self.file_id),
                message,
            )
            .into()),
            None => Err(EndOfFile.into()),
        }
    }

    pub fn eat_while<F>(&mut self, mut predicate: F)
    where
        F: FnMut(char) -> bool,
    {
        self.eat_windowed_while(|_, current| predicate(current))
    }

    pub fn eat_windowed_while<F>(&mut self, mut predicate: F)
    where
        F: FnMut(char, char) -> bool,
    {
        let mut last: char = '\0';
        while let Some(char) = self.peek() {
            if !predicate(last, char) {
                break;
            }

            last = char;
            self.try_consume();
        }
    }
}
