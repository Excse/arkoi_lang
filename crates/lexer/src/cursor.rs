use std::iter::Peekable;
use std::str::Chars;

use diagnostics::{SourceDetails, Span};

use crate::{reports, LexerError};

pub struct Cursor<'a> {
    source_details: &'a SourceDetails,
    chars: Peekable<Chars<'a>>,
    position: usize,
    start: usize,
    line: usize,
}

impl<'a> Cursor<'a> {
    pub fn new(source_details: &'a SourceDetails) -> Cursor<'a> {
        Cursor {
            source_details,
            chars: source_details.source.chars().peekable(),
            position: 0,
            start: 0,
            line: 0,
        }
    }

    pub fn mark_start(&mut self) {
        self.start = self.position;
    }

    pub fn as_span(&mut self) -> Span<'a> {
        Span::new(self.source_details, self.line, self.start, self.position)
    }

    pub fn as_str(&self) -> &'a str {
        &self.source_details.source[self.start..self.position]
    }

    pub fn peek(&mut self) -> Option<char> {
        self.chars.peek().map(|char| *char)
    }

    pub fn consume(&mut self) -> Option<char> {
        let char = self.chars.next()?;
        if char == '\n' {
            self.line += 1;
        }

        self.position += 1;

        Some(char)
    }

    pub fn eat(&mut self, expected: char) -> Result<char, LexerError<'a>> {
        match self.peek() {
            Some(char) if char == expected => Ok(self.consume().unwrap()),
            Some(char) => reports::didnt_expect(
                char,
                Span::new(self.source_details, self.line, self.position, self.position),
                char.to_string(),
            ),
            None => Err(LexerError::EndOfFile),
        }
    }

    pub fn eat_if<F>(&mut self, predicate: F, message: &'static str) -> Result<char, LexerError<'a>>
    where
        F: FnOnce(char) -> bool,
    {
        match self.peek() {
            Some(char) if predicate(char) => Ok(self.consume().unwrap()),
            Some(char) => reports::didnt_expect(
                char,
                Span::new(self.source_details, self.line, self.position, self.position),
                message.to_string(),
            ),
            None => Err(LexerError::EndOfFile),
        }
    }

    pub fn eat_while<F>(&mut self, mut predicate: F)
    where
        F: FnMut(char) -> bool,
    {
        self.eat_windowed_while(|_, current| predicate(current))
    }

    pub fn eat_windowed_while<F>(&mut self, mut predicate: F)
    where
        F: FnMut(char, char) -> bool,
    {
        let mut last: char = '\0';
        while let Some(char) = self.peek() {
            if !predicate(last, char) {
                break;
            }

            last = char;
            self.consume();
        }
    }
}
