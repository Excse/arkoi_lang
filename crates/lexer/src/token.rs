use serde::Serialize;
use serdebug::SerDebug;
use strum::AsRefStr;

use diagnostics::Span;

#[derive(SerDebug, Serialize)]
pub struct Token<'a> {
    pub span: Span<'a>,
    pub value: Option<TokenValue<'a>>,
    pub kind: TokenKind,
}

#[derive(SerDebug, Serialize)]
pub enum TokenValue<'a> {
    Integer(usize),
    Decimal(f64),
    String(&'a str),
    Boolean(bool),
}

#[derive(SerDebug, Serialize, Eq, PartialEq, Copy, Clone, AsRefStr)]
pub enum TokenKind {
    #[serde(rename = "int")]
    #[strum(serialize = "int")]
    Integer,
    #[serde(rename = "decimal")]
    #[strum(serialize = "decimal")]
    Decimal,
    #[serde(rename = "identifier")]
    #[strum(serialize = "identifier")]
    Identifier,
    #[serde(rename = "string")]
    #[strum(serialize = "string")]
    String,
    #[serde(rename = "true")]
    #[strum(serialize = "true")]
    True,
    #[serde(rename = "false")]
    #[strum(serialize = "false")]
    False,

    #[serde(rename = "struct")]
    #[strum(serialize = "struct")]
    Struct,
    #[serde(rename = "fun")]
    #[strum(serialize = "fun")]
    Fun,
    #[serde(rename = "let")]
    #[strum(serialize = "let")]
    Let,
    #[serde(rename = "return")]
    #[strum(serialize = "return")]
    Return,

    #[serde(rename = "{")]
    #[strum(serialize = "{")]
    OBracket,
    #[serde(rename = "}")]
    #[strum(serialize = "}")]
    CBracket,
    #[serde(rename = "(")]
    #[strum(serialize = "(")]
    OParent,
    #[serde(rename = ")")]
    #[strum(serialize = ")")]
    CParent,
    #[serde(rename = "@")]
    #[strum(serialize = "@")]
    At,
    #[serde(rename = "!")]
    #[strum(serialize = "|")]
    Apostrophe,
    #[serde(rename = ",")]
    #[strum(serialize = ",")]
    Comma,
    #[serde(rename = ".")]
    #[strum(serialize = ".")]
    Period,
    #[serde(rename = ";")]
    #[strum(serialize = ";")]
    Semicolon,

    #[serde(rename = "+=")]
    #[strum(serialize = "+=")]
    AddAssign,
    #[serde(rename = "+")]
    #[strum(serialize = "+=")]
    Plus,
    #[serde(rename = "-=")]
    #[strum(serialize = "-=")]
    MinusAssign,
    #[serde(rename = "-")]
    #[strum(serialize = "-")]
    Minus,
    #[serde(rename = "*=")]
    #[strum(serialize = "*=")]
    MulAssign,
    #[serde(rename = "*")]
    #[strum(serialize = "*")]
    Asterisk,
    #[serde(rename = "/=")]
    #[strum(serialize = "/=")]
    DivAssign,
    #[serde(rename = "/")]
    #[strum(serialize = "/")]
    Slash,
    #[serde(rename = "<=")]
    #[strum(serialize = "<=")]
    LessEqual,
    #[serde(rename = "<")]
    #[strum(serialize = "<")]
    Less,
    #[serde(rename = ">=")]
    #[strum(serialize = ">=")]
    GreaterEqual,
    #[serde(rename = ">")]
    #[strum(serialize = ">")]
    Greater,
    #[serde(rename = "==")]
    #[strum(serialize = "==")]
    Equal,
    #[serde(rename = "!=")]
    #[strum(serialize = "!=")]
    NotEqual,
    #[serde(rename = "=")]
    #[strum(serialize = "=")]
    Assign,

    #[serde(rename = "self")]
    #[strum(serialize = "self")]
    Self_,
    #[serde(rename = "u8")]
    #[strum(serialize = "u8")]
    U8,
    #[serde(rename = "i8")]
    #[strum(serialize = "i8")]
    I8,
    #[serde(rename = "u16")]
    #[strum(serialize = "u16")]
    U16,
    #[serde(rename = "i16")]
    #[strum(serialize = "i16")]
    I16,
    #[serde(rename = "u32")]
    #[strum(serialize = "u32")]
    U32,
    #[serde(rename = "i32")]
    #[strum(serialize = "i32")]
    I32,
    #[serde(rename = "u64")]
    #[strum(serialize = "u64")]
    U64,
    #[serde(rename = "i64")]
    #[strum(serialize = "i64")]
    I64,
    #[serde(rename = "usize")]
    #[strum(serialize = "usize")]
    USize,
    #[serde(rename = "isize")]
    #[strum(serialize = "isize")]
    ISize,
    #[serde(rename = "f32")]
    #[strum(serialize = "f32")]
    F32,
    #[serde(rename = "f64")]
    #[strum(serialize = "f64")]
    F64,

    #[serde(rename = "unknown")]
    #[strum(serialize = "unknown")]
    Unknown,
}

impl<'a> Token<'a> {
    pub fn new(span: Span<'a>, value: Option<TokenValue<'a>>, kind: TokenKind) -> Token<'a> {
        Token { span, value, kind }
    }
}
