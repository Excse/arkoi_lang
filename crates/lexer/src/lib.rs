pub mod cursor;
pub mod token;

use diagnostics::SourceDetails;
use serdebug::SerDebug;
use serde::Serialize;
use token::{Token, TokenKind};
use cursor::Cursor;

pub struct Lexer<'a> {
    cursor: Cursor<'a>,
}

#[derive(SerDebug, Serialize)]
pub enum LexerError {
    DidntExpect(char, &'static str),
    InternalError(&'static str),
    EndOfFile,
}

pub struct TokenIter<'a> {
    lexer: &'a mut Lexer<'a>,
}

impl<'a> Iterator for TokenIter<'a> {
    type Item = Token<'a>;

    fn next(&mut self) -> Option<Self::Item> {
        match self.lexer.next_token() {
            Ok(token_kind) => Some(Token::new(self.lexer.cursor.as_span(), token_kind)),
            Err(_) => None,
        }
    }
}

impl<'a> Lexer<'a> {
    pub fn new(source_details: &'a SourceDetails) -> Lexer<'a> {
        Lexer {
            cursor: Cursor::new(source_details),
        }
    }

    pub fn iter(&'a mut self) -> TokenIter<'a> {
        TokenIter { lexer: self }
    }

    fn next_token(&mut self) -> Result<TokenKind<'a>, LexerError> {
        self.cursor.mark_start();
        match self.cursor.peek() {
            Some(char) if char.is_alphabetic() => self.read_identifier(),
            Some(char) if char.is_numeric() => self.read_number(),
            Some('"') => self.read_string(),
            Some(_) => self.read_symbol(),
            None => Err(LexerError::EndOfFile),
        }
    }

    fn read_symbol(&mut self) -> Result<TokenKind<'a>, LexerError> {
        let mut token = match self.cursor.consume() {
            Some(char) if char.is_whitespace() => self.next_token()?,
            Some('{') => TokenKind::OBracket,
            Some('}') => TokenKind::CBracket,
            Some('(') => TokenKind::OParent,
            Some(')') => TokenKind::CParent,
            Some('@') => TokenKind::At,
            Some(',') => TokenKind::Comma,
            Some('.') => TokenKind::Period,
            Some('+') => TokenKind::Plus,
            Some('-') => TokenKind::Minus,
            Some('*') => TokenKind::Asterisk,
            Some('/') => TokenKind::Slash,
            Some('<') => TokenKind::Less,
            Some('>') => TokenKind::Greater,
            Some('=') => TokenKind::Assign,
            Some('!') => TokenKind::Apostrophe,
            Some(';') => TokenKind::Semicolon,
            _ => TokenKind::Unknown,
        };

        let current = match self.cursor.peek() {
            Some(char) => char,
            None => return Ok(token),
        };

        token = match (token, current) {
            (TokenKind::Plus, '=') => TokenKind::AddAssign,
            (TokenKind::Minus, '=') => TokenKind::MinusAssign,
            (TokenKind::Asterisk, '=') => TokenKind::MulAssign,
            (TokenKind::Slash, '=') => TokenKind::DivAssign,
            (TokenKind::Less, '=') => TokenKind::LessEqual,
            (TokenKind::Greater, '=') => TokenKind::GreaterEqual,
            (TokenKind::Assign, '=') => TokenKind::Equal,
            (TokenKind::Apostrophe, '=') => TokenKind::NotEqual,
            (token, _) => return Ok(token),
        };

        self.cursor.consume();

        Ok(token)
    }

    fn read_identifier(&mut self) -> Result<TokenKind<'a>, LexerError> {
        match self.cursor.consume() {
            Some(char) if !char.is_alphabetic() => {
                return Err(LexerError::DidntExpect(char, "a-zA-Z"))
            }
            Some(_) => {}
            None => return Err(LexerError::EndOfFile),
        }

        self.cursor.eat_while(char::is_alphanumeric);

        let identifier_name = self.cursor.as_str();
        Ok(match identifier_name {
            "true" => TokenKind::Boolean(true),
            "false" => TokenKind::Boolean(false),
            "struct" => TokenKind::Struct,
            "return" => TokenKind::Return,
            "let" => TokenKind::Let,
            "self" => TokenKind::Self_,
            "fun" => TokenKind::Fun,
            "u8" => TokenKind::U8,
            "i8" => TokenKind::I8,
            "u16" => TokenKind::U16,
            "i16" => TokenKind::I16,
            "u32" => TokenKind::U32,
            "i32" => TokenKind::I32,
            "u64" => TokenKind::U64,
            "i64" => TokenKind::I64,
            "usize" => TokenKind::USize,
            "isize" => TokenKind::ISize,
            "f32" => TokenKind::F32,
            "f64" => TokenKind::F64,
            _ => TokenKind::Identifier(identifier_name),
        })
    }

    fn read_number(&mut self) -> Result<TokenKind<'a>, LexerError> {
        match self.cursor.consume() {
            Some(char) if char.is_numeric() => {}
            Some(char) => return Err(LexerError::DidntExpect(char, "0-9")),
            None => return Err(LexerError::EndOfFile),
        }

        self.cursor.eat_while(char::is_numeric);

        match self.cursor.peek() {
            Some('.') => {
                self.cursor.consume();
                self.cursor.eat_while(char::is_numeric);

                let number = self.cursor.as_str();
                number
                    .parse::<f64>()
                    .map(TokenKind::Decimal)
                    .map_err(|_| LexerError::InternalError("Couldn't parse the string to a f64."))
            }
            _ => {
                let number = self.cursor.as_str();
                number
                    .parse::<usize>()
                    .map(TokenKind::Integer)
                    .map_err(|_| LexerError::InternalError("Couldn't parse the string to a usize."))
            }
        }
    }

    fn read_string(&mut self) -> Result<TokenKind<'a>, LexerError> {
        match self.cursor.consume() {
            Some('"') => {}
            Some(char) => return Err(LexerError::DidntExpect(char, "\"")),
            None => return Err(LexerError::EndOfFile),
        };

        self.cursor
            .eat_windowed_while(|prev, curr| curr != '"' || prev == '\\');

        match self.cursor.consume() {
            Some('"') => {}
            Some(char) => return Err(LexerError::DidntExpect(char, "\"")),
            None => return Err(LexerError::EndOfFile),
        };

        let string_content = self.cursor.as_str();
        let string_content = &string_content[1..string_content.len() - 1];
        Ok(TokenKind::String(string_content))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    macro_rules! test_token {
        (FAIL: $name:ident, $func:ident, $source:expr) => {
            #[test]
            fn $name() {
                let source_details = SourceDetails::new($source, "test.ark");
                let mut lexer = Lexer::new(&source_details);
                let token = lexer.$func();
                assert!(token.is_err(), "{:?} should be an error", token);
            }
        };
        ($name:ident, $source:expr => $expected:expr) => {
            #[test]
            fn $name() {
                let source_details = SourceDetails::new($source, "test.ark");
                let mut lexer = Lexer::new(&source_details);
                let expected = TokenKind::from($expected);
                let token = lexer.next_token().unwrap();
                assert!(token.same_variant(&expected), "Input was {:?}", $source);
            }
        };
    }

    test_token!(success_decimal, "4.2" => 4.2);
    test_token!(success_integer, "42" => 42);
    test_token!(FAIL: fail_number, read_number, "number");

    test_token!(success_string, "\"Hello World!\"" => TokenKind::String("Hello World!"));
    test_token!(FAIL: fail_string, read_string, "Hello World!");

    test_token!(success_true, "true" => TokenKind::Boolean(true));
    test_token!(success_false, "false" => TokenKind::Boolean(false));

    test_token!(success_bool, "true" => TokenKind::Boolean(true));
    test_token!(success_obracket, "{" => TokenKind::OBracket);
    test_token!(success_cbracket, "}" => TokenKind::CBracket);
    test_token!(success_oparent, "(" => TokenKind::OParent);
    test_token!(success_cparent, ")" => TokenKind::CParent);
    test_token!(success_at, "@" => TokenKind::At);
    test_token!(success_apostrophe, "!" => TokenKind::Apostrophe);
    test_token!(success_comma, "," => TokenKind::Comma);
    test_token!(success_period, "." => TokenKind::Period);
    test_token!(success_semicolon, ";" => TokenKind::Semicolon);
    test_token!(success_addassign, "+=" => TokenKind::AddAssign);
    test_token!(success_plus, "+" => TokenKind::Plus);
    test_token!(success_minusassing, "-=" => TokenKind::MinusAssign);
    test_token!(success_minus, "-" => TokenKind::Minus);
    test_token!(success_mulassign, "*=" => TokenKind::MulAssign);
    test_token!(success_asterisk, "*" => TokenKind::Asterisk);
    test_token!(success_divassign, "/=" => TokenKind::DivAssign);
    test_token!(success_slash, "/" => TokenKind::Slash);
    test_token!(success_lessequal, "<=" => TokenKind::LessEqual);
    test_token!(success_less, "<" => TokenKind::Less);
    test_token!(success_greaterequal, ">=" => TokenKind::GreaterEqual);
    test_token!(success_greater, ">" => TokenKind::Greater);
    test_token!(success_equal, "==" => TokenKind::Equal);
    test_token!(success_notequal, "!=" => TokenKind::NotEqual);
    test_token!(success_assign, "=" => TokenKind::Assign);

    test_token!(success_self, "self" => TokenKind::Self_);
    test_token!(success_u8, "u8" => TokenKind::U8);
    test_token!(success_i8, "i8" => TokenKind::I8);
    test_token!(success_u16, "u16" => TokenKind::U16);
    test_token!(success_i16, "i16" => TokenKind::I16);
    test_token!(success_u32, "u32" => TokenKind::U32);
    test_token!(success_i32, "i32" => TokenKind::I32);
    test_token!(success_u64, "u64" => TokenKind::U64);
    test_token!(success_i64, "i64" => TokenKind::I64);
    test_token!(success_usize, "usize" => TokenKind::USize);
    test_token!(success_isize, "isize" => TokenKind::ISize);
    test_token!(success_f32, "f32" => TokenKind::F32);
    test_token!(success_f64, "f64" => TokenKind::F64);

    macro_rules! insta_test {
        ($name:ident, $path:expr) => {
            #[test]
            fn $name() {
                let source_details = match SourceDetails::read($path) {
                    Ok(source_details) => source_details,
                    Err(err) => panic!("{err}"),
                };

                let mut lexer = Lexer::new(&source_details);
                let tokens = lexer.iter().collect::<Vec<Token>>();

                insta::assert_yaml_snapshot!(&tokens);
            }
        };
    }

    insta_test!(insta_test, "test_files/insta_test.ark");
}
