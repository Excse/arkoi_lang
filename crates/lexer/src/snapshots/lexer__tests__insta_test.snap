---
source: crates/lexer/src/lib.rs
expression: "&tokens"
---
- span:
    line: 0
    start: 0
    end: 6
  value: ~
  kind: struct
- span:
    line: 0
    start: 7
    end: 10
  value:
    String: Foo
  kind: identifier
- span:
    line: 0
    start: 11
    end: 12
  value: ~
  kind: "{"
- span:
    line: 1
    start: 14
    end: 17
  value:
    String: foo
  kind: identifier
- span:
    line: 1
    start: 18
    end: 19
  value: ~
  kind: "@"
- span:
    line: 1
    start: 19
    end: 22
  value: ~
  kind: i32
- span:
    line: 1
    start: 22
    end: 23
  value: ~
  kind: "!"
- span:
    line: 1
    start: 23
    end: 24
  value: ~
  kind: ","
- span:
    line: 2
    start: 26
    end: 29
  value:
    String: bar
  kind: identifier
- span:
    line: 2
    start: 30
    end: 31
  value: ~
  kind: "@"
- span:
    line: 2
    start: 31
    end: 34
  value: ~
  kind: i64
- span:
    line: 2
    start: 34
    end: 35
  value: ~
  kind: ","
- span:
    line: 3
    start: 36
    end: 37
  value: ~
  kind: "}"
- span:
    line: 5
    start: 39
    end: 42
  value: ~
  kind: fun
- span:
    line: 5
    start: 43
    end: 47
  value:
    String: init
  kind: identifier
- span:
    line: 5
    start: 47
    end: 48
  value: ~
  kind: (
- span:
    line: 5
    start: 48
    end: 52
  value: ~
  kind: self
- span:
    line: 5
    start: 53
    end: 54
  value: ~
  kind: "@"
- span:
    line: 5
    start: 54
    end: 57
  value:
    String: Foo
  kind: identifier
- span:
    line: 5
    start: 57
    end: 58
  value: ~
  kind: "!"
- span:
    line: 5
    start: 58
    end: 59
  value: ~
  kind: )
- span:
    line: 5
    start: 60
    end: 61
  value: ~
  kind: "@"
- span:
    line: 5
    start: 61
    end: 65
  value:
    String: void
  kind: identifier
- span:
    line: 5
    start: 66
    end: 67
  value: ~
  kind: "{"
- span:
    line: 6
    start: 69
    end: 73
  value: ~
  kind: self
- span:
    line: 6
    start: 73
    end: 74
  value: ~
  kind: "."
- span:
    line: 6
    start: 74
    end: 77
  value:
    String: foo
  kind: identifier
- span:
    line: 6
    start: 78
    end: 79
  value: ~
  kind: "="
- span:
    line: 6
    start: 80
    end: 82
  value:
    Integer: 42
  kind: int
- span:
    line: 6
    start: 82
    end: 83
  value: ~
  kind: ;
- span:
    line: 7
    start: 84
    end: 85
  value: ~
  kind: "}"
- span:
    line: 9
    start: 87
    end: 93
  value: ~
  kind: struct
- span:
    line: 9
    start: 94
    end: 98
  value:
    String: Test
  kind: identifier
- span:
    line: 9
    start: 99
    end: 100
  value: ~
  kind: "{"
- span:
    line: 10
    start: 103
    end: 108
  value:
    String: hello
  kind: identifier
- span:
    line: 10
    start: 109
    end: 110
  value: ~
  kind: "@"
- span:
    line: 10
    start: 110
    end: 113
  value: ~
  kind: i64
- span:
    line: 10
    start: 113
    end: 114
  value: ~
  kind: "!"
- span:
    line: 10
    start: 114
    end: 115
  value: ~
  kind: ","
- span:
    line: 11
    start: 117
    end: 122
  value:
    String: world
  kind: identifier
- span:
    line: 11
    start: 123
    end: 124
  value: ~
  kind: "@"
- span:
    line: 11
    start: 124
    end: 127
  value: ~
  kind: i64
- span:
    line: 11
    start: 127
    end: 128
  value: ~
  kind: "!"
- span:
    line: 11
    start: 128
    end: 129
  value: ~
  kind: ","
- span:
    line: 12
    start: 131
    end: 134
  value:
    String: foo
  kind: identifier
- span:
    line: 12
    start: 135
    end: 136
  value: ~
  kind: "@"
- span:
    line: 12
    start: 136
    end: 139
  value:
    String: Foo
  kind: identifier
- span:
    line: 12
    start: 139
    end: 140
  value: ~
  kind: ","
- span:
    line: 13
    start: 141
    end: 142
  value: ~
  kind: "}"
- span:
    line: 15
    start: 144
    end: 147
  value: ~
  kind: fun
- span:
    line: 15
    start: 148
    end: 152
  value:
    String: init
  kind: identifier
- span:
    line: 15
    start: 152
    end: 153
  value: ~
  kind: (
- span:
    line: 15
    start: 153
    end: 157
  value: ~
  kind: self
- span:
    line: 15
    start: 158
    end: 159
  value: ~
  kind: "@"
- span:
    line: 15
    start: 159
    end: 163
  value:
    String: Test
  kind: identifier
- span:
    line: 15
    start: 163
    end: 164
  value: ~
  kind: "!"
- span:
    line: 15
    start: 164
    end: 165
  value: ~
  kind: )
- span:
    line: 15
    start: 166
    end: 167
  value: ~
  kind: "@"
- span:
    line: 15
    start: 167
    end: 171
  value:
    String: void
  kind: identifier
- span:
    line: 15
    start: 172
    end: 173
  value: ~
  kind: "{"
- span:
    line: 16
    start: 175
    end: 179
  value: ~
  kind: self
- span:
    line: 16
    start: 179
    end: 180
  value: ~
  kind: "."
- span:
    line: 16
    start: 180
    end: 185
  value:
    String: hello
  kind: identifier
- span:
    line: 16
    start: 186
    end: 187
  value: ~
  kind: "="
- span:
    line: 16
    start: 188
    end: 190
  value:
    Integer: 21
  kind: int
- span:
    line: 16
    start: 190
    end: 191
  value: ~
  kind: ;
- span:
    line: 17
    start: 193
    end: 197
  value: ~
  kind: self
- span:
    line: 17
    start: 197
    end: 198
  value: ~
  kind: "."
- span:
    line: 17
    start: 198
    end: 203
  value:
    String: world
  kind: identifier
- span:
    line: 17
    start: 204
    end: 205
  value: ~
  kind: "="
- span:
    line: 17
    start: 206
    end: 208
  value:
    Integer: 21
  kind: int
- span:
    line: 17
    start: 208
    end: 209
  value: ~
  kind: ;
- span:
    line: 18
    start: 210
    end: 211
  value: ~
  kind: "}"
- span:
    line: 20
    start: 213
    end: 216
  value: ~
  kind: fun
- span:
    line: 20
    start: 217
    end: 220
  value:
    String: add
  kind: identifier
- span:
    line: 20
    start: 220
    end: 221
  value: ~
  kind: (
- span:
    line: 20
    start: 221
    end: 225
  value: ~
  kind: self
- span:
    line: 20
    start: 226
    end: 227
  value: ~
  kind: "@"
- span:
    line: 20
    start: 227
    end: 231
  value:
    String: Test
  kind: identifier
- span:
    line: 20
    start: 231
    end: 232
  value: ~
  kind: )
- span:
    line: 20
    start: 233
    end: 234
  value: ~
  kind: "@"
- span:
    line: 20
    start: 234
    end: 237
  value: ~
  kind: i64
- span:
    line: 20
    start: 238
    end: 239
  value: ~
  kind: "{"
- span:
    line: 21
    start: 241
    end: 245
  value: ~
  kind: self
- span:
    line: 21
    start: 245
    end: 246
  value: ~
  kind: "."
- span:
    line: 21
    start: 246
    end: 251
  value:
    String: hello
  kind: identifier
- span:
    line: 21
    start: 252
    end: 253
  value: ~
  kind: +
- span:
    line: 21
    start: 254
    end: 258
  value: ~
  kind: self
- span:
    line: 21
    start: 258
    end: 259
  value: ~
  kind: "."
- span:
    line: 21
    start: 259
    end: 264
  value:
    String: world
  kind: identifier
- span:
    line: 21
    start: 265
    end: 266
  value: ~
  kind: +
- span:
    line: 21
    start: 267
    end: 271
  value: ~
  kind: self
- span:
    line: 21
    start: 271
    end: 272
  value: ~
  kind: "."
- span:
    line: 21
    start: 272
    end: 275
  value:
    String: foo
  kind: identifier
- span:
    line: 21
    start: 275
    end: 276
  value: ~
  kind: "."
- span:
    line: 21
    start: 276
    end: 279
  value:
    String: foo
  kind: identifier
- span:
    line: 21
    start: 280
    end: 281
  value: ~
  kind: +
- span:
    line: 21
    start: 282
    end: 286
  value: ~
  kind: self
- span:
    line: 21
    start: 286
    end: 287
  value: ~
  kind: "."
- span:
    line: 21
    start: 287
    end: 290
  value:
    String: foo
  kind: identifier
- span:
    line: 21
    start: 290
    end: 291
  value: ~
  kind: "."
- span:
    line: 21
    start: 291
    end: 294
  value:
    String: bar
  kind: identifier
- span:
    line: 22
    start: 295
    end: 296
  value: ~
  kind: "}"
- span:
    line: 24
    start: 298
    end: 301
  value: ~
  kind: fun
- span:
    line: 24
    start: 302
    end: 306
  value:
    String: main
  kind: identifier
- span:
    line: 24
    start: 306
    end: 307
  value: ~
  kind: (
- span:
    line: 24
    start: 307
    end: 308
  value: ~
  kind: )
- span:
    line: 24
    start: 309
    end: 310
  value: ~
  kind: "@"
- span:
    line: 24
    start: 310
    end: 313
  value: ~
  kind: i32
- span:
    line: 24
    start: 314
    end: 315
  value: ~
  kind: "{"
- span:
    line: 25
    start: 317
    end: 320
  value: ~
  kind: let
- span:
    line: 25
    start: 321
    end: 325
  value:
    String: test
  kind: identifier
- span:
    line: 25
    start: 326
    end: 327
  value: ~
  kind: "@"
- span:
    line: 25
    start: 327
    end: 330
  value: ~
  kind: i32
- span:
    line: 25
    start: 331
    end: 332
  value: ~
  kind: "="
- span:
    line: 25
    start: 333
    end: 334
  value:
    Integer: 0
  kind: int
- span:
    line: 25
    start: 334
    end: 335
  value: ~
  kind: ;
- span:
    line: 26
    start: 337
    end: 341
  value:
    String: test
  kind: identifier
- span:
    line: 27
    start: 342
    end: 343
  value: ~
  kind: "}"

